File: agents/intervention_agent.py
from agents.state import GraphState


class InterventionAgentNode:
    """LangGraph node that selects the best intervention strategy."""

    def __call__(self, state: GraphState) -> dict:
        # 1. Retrieve data from State
        diagnosis = state["diagnosis"]
        kb = state["knowledge_base"]

        concept = diagnosis.get("weakest_concept")
        error_tag = diagnosis.get("top_error_tag")
        misconception = diagnosis.get("misconception")

        # 2. Find matching interventions from knowledge base
        matches = []
        for item in kb.get("interventions", {}).get("rules", []):
            if item["concept_id"] != concept:
                continue
            if misconception and item["misconception_id"] == misconception["id"]:
                matches.append(item)

        selected = matches[0] if matches else {
            "concept": concept,
            "error_tag": error_tag,
            "strategy": "Targeted review",
            "activities": ["Review the concept with one worked example and one independent problem."],
            "why": "Fallback strategy when no exact intervention rule exists.",
        }

        # 3. Return the key updates to the GraphState
        return {
            "selected_intervention": selected,
            "trace": state["trace"] + [{"agent": "intervention-agent", "message": f"Selected intervention strategy '{selected.get('strategy', 'N/A')}'."}],
        }

File: backend-node/src/agents/interventionAgent.js
import { invokeAgentStructured } from "../services/llmClient.js";
import { INTERVENTION_PROMPT } from "./prompts.js";
import { InterventionOutputSchema } from "./agentSchemas.js";

/**
 * Selects the best intervention strategy using an LLM.
 */
export async function intervene(state) {
  const { diagnosis, knowledge_base: kb, prior_interventions = [], retrieval_context = {} } = state;

  const concept = diagnosis.weakest_concept;
  const error_tag = diagnosis.top_error_tag;
  const misconception = diagnosis.misconception;

  // filter rules by concept
  const available_rules = (kb.interventions?.rules || []).filter(
    (r) => r.concept_id === concept
  );

  const context = {
    diagnosed_concept: concept,
    top_error_tag: error_tag,
    misconception,
    available_rules,
    graph_context: retrieval_context.graph_context || [],
    prior_interventions,
  };

  const userContent = `Context:\n${JSON.stringify(context, null, 2)}`;

  const result = await invokeAgentStructured({
    systemPrompt: INTERVENTION_PROMPT,
    userContent,
    responseSchema: InterventionOutputSchema,
  });

  state.selected_intervention = {
    concept,
    error_tag,
    strategy: result.strategy,
    activities: result.activities,
    why: result.why,
    estimated_sessions: result.estimated_sessions || 2,
  };
  state.trace.push({
    agent: "intervention-agent",
    message: `Selected intervention strategy '${result.strategy}'.`,
  });

  return state;
}
